langroid/agent/base.py

```diff
@@ -594,10 +610,65 @@ def handle_tool_message(self, tool: ToolMessage) -> None | str | ChatDocument:
             result = f"Error in tool/function-call {tool_name} usage: {type(e)}: {e}"
         return result  # type: ignore
 
-    def num_tokens(self, prompt: str) -> int:
+    def num_tokens(self, prompt: str | List[LLMMessage]) -> int:
         if self.parser is None:
             raise ValueError("Parser must be set, to count tokens")
-        return self.parser.num_tokens(prompt)
+        if isinstance(prompt, str):
+            return self.parser.num_tokens(prompt)
+        else:
+            return sum([self.parser.num_tokens(m.content) for m in prompt])
+
+    def update_token_usage(
+        self, response: LLMResponse, prompt: str | List[LLMMessage], stream: bool
+    ) -> None:
+        """
+        Updates the usage memebr field in the response.
+        It updates the cost after checking the cache and updates the
+        tokens (prompts and completion) if the response stream is True, because OpenAI
+        doesn't returns these fields.
+
+        Args:
+            response (LLMResponse): LLMResponse object
+            prompt (str | List[LLMMessage]): prompt or list of LLMMessage objects
+            stream (bool): whether to update the usage in the response object
+                if the response is not cached.
+        """
+        if response is not None:
+            if stream:
+                prompt_tokens = self.num_tokens(prompt)
```

Comment:
when response is from cache, shouldn't we set the token usages to 0 ?

Reply:
In general, we want the `LLMTokenUsage` fields to reflect non-cached usage,  and actual cost.

Reply:
> when response is from cache, shouldn't we set the token usages to 0 ?

I my interpretation is that these responses are ultimately LLM tokens, so we need to count them, but with NO implication on the cost. But I agree, the notion of reflecting **non-cached usage** is cleaner.